applications/luci-app-openvpn/luasrc/controller/openvpn.lua

```diff
@@ -8,4 +8,47 @@ function index()
 	entry( {"admin", "services", "openvpn"}, cbi("openvpn"), _("OpenVPN") )
 	entry( {"admin", "services", "openvpn", "basic"},    cbi("openvpn-basic"),    nil ).leaf = true
 	entry( {"admin", "services", "openvpn", "advanced"}, cbi("openvpn-advanced"), nil ).leaf = true
+	entry( {"admin", "services", "openvpn", "file"},     form("openvpn-file"),    nil ).leaf = true
+	entry( {"admin", "services", "openvpn", "upload"},   call("ovpn_upload"))
+end
+
+function ovpn_upload()
+	local fs     = require("nixio.fs")
+	local http   = require("luci.http")
+	local util   = require("luci.util")
+	local uci    = require("luci.model.uci").cursor()
+	local upload = http.formvalue("ovpn_file")
+	local name   = http.formvalue("instance_name2")
+	local file   = "/etc/openvpn/" ..name.. ".ovpn"
+
+	if name and upload then
+		local fp
+
+		http.setfilehandler(
+			function(meta, chunk, eof)
+				local data = util.trim(chunk:gsub("\r\n", "\n")) .. "\n"
+				data = util.trim(data:gsub("[\128-\255]", ""))
+
+				if not fp and meta and meta.name == "ovpn_file" then
+					fp = io.open(file, "w")
+				end
+				if fp and data then
+					fp:write(data)
+				end
+				if fp and eof then
+					fp:close()
+				end
+			end
+		)
+
+		if fs.access(file) then
+			if not uci:get_first("openvpn", name) then
```

Comment:
If the section already exists but points to another file, the currently uploaded configuration will be orphaned and unused, maybe unlink it then to avoid piling up garbage?

Something along the lines of

```
    ...
elseif uci:get("openvpn", name, "file") ~= file then
    fs.unlink(file)
end
```

Reply:
see above, you can't upload/overwrite an existing section.